### DISCUSSION TITLE
Is a static constructor to configure spdlog safe to use?

### DISCUSSION THREAD

**Discussion by CPickens42** [gabime/spdlog#2362] [Q&A] [answered]
Hello,

I'm using spdlog and I have this in my code (C++20):

```C++
#include <spdlog/spdlog.h>
#include <spdlog/sinks/stdout_color_sinks.h>


[[maybe_unused]] auto static_ctor = std::invoke([]() {  /* NOLINT */
    // see https://spdlog.docsforge.com/v1.x/0.faq/#switch-the-default-logger-to-stderr
    spdlog::set_default_logger(spdlog::stderr_color_mt("some_arbitrary_name"));
    spdlog::set_default_logger(spdlog::stderr_color_mt(""));
    spdlog::set_pattern("%^[%Y-%m-%d %T.%F | %-8l]%$ %v");
    return nullptr;
});
```

I'm unsure of the safety of this. Could I encounter a ["static initialization fiasco"](https://isocpp.org/wiki/faq/ctors#static-init-order)? If so, in which cases?
So far I don't see the issue since, for what I saw, there is no static initialization in spdlog, but I would like to know a bit better.

This is done like this to avoid to repeat the configuration of spdlog in every tests, in every main and so on.

**Comment by Iravoo** [comment, marked as answer]
I'm interested in this as well. Maybe @tt4g has an idea? :slightly_smiling_face:

**Reply by tt4g** [reply]
In my personal opinion, I do not use `static` variables to avoid "Don't repeat yourself".
It may crash depending on the initialization order of `static` variables as described in [static initialization fiasco](https://isocpp.org/wiki/faq/ctors#static-init-order) And the initialization order of `static` variables is difficult for developers to control.
It is a real shame to struggle with `static` variables introduced for the sake of ease.

Note that `static` variables can cause problems depending on the order in which the destructors are called.

See also:

- #2113
- #2298

I would define the functions that configure spdlog.
Calling the functions that configure spdlog repeatedly in `main()` is tedious, but the tedious work is better than the risk of encountering bugs that cannot be avoided.

**Reply by CPickens42** [reply]
Thank you @tt4g, your answer was really clear. :slightly_smiling_face: !
